### DISCUSSION TITLE
Is there a way to tell if a PR is from a forked repository?

### DISCUSSION THREAD

**Discussion by Senui** [community/community#26829] [Actions]
<p>Dear Github Community,</p>
<p>At the end of each of my workflows, a Slack message (using a Slack action) is sent using <code>GITHUB_TOKEN</code> from my base repository. Now if someone wishes to make a PR from a forked repository, this Slack step will fail because <code>GITHUB_TOKEN</code> being inaccessible outside the base repository. As a result, any PR that is made from a forked repository fails.</p>
<p>My question is: can I somehow not run the Slack action step when a PR is made from a forked repository? For example:</p>
<pre><code class="lang-auto"> name: Notify Slack
      uses: 8398a7/action-slack@v3
      ...
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
        MATRIX_CONTEXT: ${{ toJson(matrix) }} # Required when using matrix
      # Pick up events if the job succeeds, fails or is canceled, and if this
      # job is not executed in forked repositories
      if: (${{ failure() }} || ${{ success() }} || ${{ cancelled() }}) &amp;&amp; ${{ github.repository.is_not_forked }}
</code></pre>
<p>I have also thought about allowing the Slack action step to fail, but there seems to be no such feature. <code>continue-on-error</code> will still end up with those red crosses for PRs from forked repos.</p>
<p>Any thoughts?</p>

**Comment by kingthorin** [comment]
<p>Assuming you do actually want it to run: This is the type of thing the <code>pull_request_target</code> event is meant to facilitate.</p>
<ul>
<li><a href="https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/" rel="noopener nofollow ugc">https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/</a></li>
<li><a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_target" rel="noopener nofollow ugc">https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_target</a></li>
</ul>

**Comment by Senui** [comment]
<p>Thanks for your reply. Do I understand correctly that if I append the <code>pull_request_target</code> event to my workflows’ <code>on</code> section, any PRs made from forked repositories will be able to use my base repository’s <code>GITHUB_TOKEN</code> (in this case, to send Slack messages)?</p>
<pre><code class="lang-auto">on:
  push:
  pull_request:
    branches:
      - master
  pull_request_target:
    branches:
      - master

jobs:
  build:
    runs-on: ubuntu-18.04

    steps:
    - name: Notify Slack
      uses: 8398a7/action-slack@v3
      ...
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
        MATRIX_CONTEXT: ${{ toJson(matrix) }} # Required when using matrix
      if: always()


</code></pre>
<p>Would I still need the regular <code>pull_request</code> event in this case for PRs made from the base repository?</p>

**Comment by kingthorin** [comment]
<p>As far as I understand you should be able to use <code>pull_request_target</code> most places your previously used <code>pull_request</code> if it needs access to tokens (it’s like the original + :slight_smile:).</p>
<p>Depending what your workflow does you may need to also ensure you’ve checked out the proper code to analyze/build etc.</p>
<pre><code class="lang-auto">      uses: actions/checkout@v2
      with:
        ref: ${{github.event.pull_request.head.ref}}
        repository: ${{github.event.pull_request.head.repo.full_name}}
</code></pre>

**Reply by koczkatamas** [reply]
Unrelated note: be warned that if you use this code in a public repository with `pull_request_target` trigger, then by default it will run the PR author's code while the available `GITHUB_TOKEN` has wide (including write) permissions, so they can steal your secrets, hijack your releases, etc. Read more here: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

**Comment by Simran-B** [comment]
<p>If <code>ref</code> and <code>repository</code> were not specified, then it would checkout the code of the target branch, correct?</p>
<blockquote>
<p>instead of running against the workflow and code from the merge commit, the event runs against the workflow and code from the base of the pull request</p>
</blockquote>

**Comment by kingthorin** [comment]
<p>Yup, that seems to be the case from my experience.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/senui">@senui</a> ,<br>
When the workflow is triggered by the events that are related to pull request ( <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request" rel="noopener nofollow ugc"><strong>pull_request</strong></a>, <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_review" rel="noopener nofollow ugc"><strong>pull_request_review</strong></a>, <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_review_comment" rel="noopener nofollow ugc"><strong>pull_request_review_comment</strong></a> or <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_target" rel="noopener nofollow ugc"><strong>pull_request_target</strong></a>), from the <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="noopener nofollow ugc">github context</a> of the workflow run, you can get the following three properties:</p>
<ul>
<li>
<p><strong>github.event.pull_request.base.repo.full_name</strong> - - The full name of PR’s <strong>base</strong> (<strong>target</strong>) repository (<strong>owner-name/repo-name</strong>).</p>
</li>
<li>
<p><strong>github.event.pull_request.head.repo.full_name</strong> - - The full name of PR’s <strong>head</strong> (<strong>source</strong>) repository (<strong>owner-name/repo-name</strong>).</p>
</li>
<li>
<p><strong>github.repository</strong> - - The full name of the repository where the PR is in. Normally this property is equal to ‘<strong>github.event.pull_request.base.repo.full_name</strong>’.</p>
</li>
</ul>
<p>When the PR’s head branch and base branch are in the same repository, the above three properties are same.<br>
When the PR’s head branch and base branch are in different repositories, ‘<strong>github.event.pull_request.head.repo.full_name</strong>’ is different with the other two.<br>
In your case, you can try this:</p>
<pre><code class="lang-yaml">steps:
  - name: Notify Slack
    if: ${{ github.event.pull_request.head.repo.full_name == github.event.pull_request.base.repo.full_name }}
    uses: 8398a7/action-slack@v3
    . . .
</code></pre>
<p>OR</p>
<pre><code class="lang-yaml">steps:
  - name: Notify Slack
    if: ${{ github.event.pull_request.head.repo.full_name == github.repository }}
    uses: 8398a7/action-slack@v3
    . . .
</code></pre>

**Comment by Senui** [comment]
<p>So I tried to go with the <code>pull_request_target</code> solution, but the main issue is that ny PRs made from the base repository to test changes to any workflow file are ignored. This makes testing of GHA workflows before merging into master not feasible.</p>
<p>I am inclined to go with <a class="mention" href="/brightran">@brightran</a>’s solution of just ignoring the Slack messages from PRs of forked repos. But I do have a question <a class="mention" href="/brightran">@brightran</a>:</p>
<pre><code class="lang-auto">steps:
 - name: Notify Slack
   if: ${{ github.event.pull_request.head.repo.full_name == github.repository }}
   uses: 8398a7/action-slack@v3
   . . .
</code></pre>
<p>Will the above step also run if a previous step failed? I used to have as condition <code>if: always()</code>, in order to also catch cases where previous steps failed or were canceled. The Slack message would then notify me of what exactly happened.</p>
<p>Would I have to make a boolean construction as I had in my opening post to take those cases into account? Something like:</p>
<pre><code class="lang-auto">if: ${{ ( failure() || success() || cancelled() ) &amp;&amp; ( github.event.pull_request.head.repo.full_name == github.event.pull_request.base.repo.full_name ) }}
</code></pre>
<p>Edit:</p>
<p>I can confirm that the above construct does indeed work: the Slack action is ignored for forked PRs and taken into account for PRs from the base repository</p>
<p>Thanks everyone!</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/senui">@senui</a>,<br>
“<strong>always()</strong>” is OK, instead of “<strong>(failure() || success() || cancelled())</strong>”. You also can reference <a href="https://github.community/t/exit-on-success/132524/2">here</a>.</p>
<pre><code class="lang-yaml">if: ${{ github.event.pull_request.head.repo.full_name == github.repository &amp;&amp; always() }}
</code></pre>

**Comment by Simran-B** [comment]
<aside class="quote no-group" data-username="senui" data-post="8" data-topic="134186">
<div class="title">
<div class="quote-controls"></div>
 senui:</div>
<blockquote>
<p>So I tried to go with the <code>pull_request_target</code> solution, but the main issue is that ny PRs made from the base repository to test changes to any workflow file are ignored. This makes testing of GHA workflows before merging into master not feasible.</p>
</blockquote>
</aside>
<p>Users without write access to your repo shouldn’t be able to execute arbitrarily modified workflows. That’s why <code>pull_request_target</code> uses the base branch’s workflows. As a project maintainer who wants to test changes to workflows, you can use <code>pull_request</code> and push a branch to the base repo instead of a fork, can’t you?</p>

**Comment by Senui** [comment]
<p><a class="mention" href="/simran-b">@Simran-B</a>, yes, but wouldn’t a PR on the base repo then trigger for both the <code>pull_request_target</code> and <code>pull_request</code> event?</p>

**Comment by Simran-B** [comment]
<p>Good point. Yes, it will trigger both if you do <code>on: [pull_request, pull_request_target]</code>.<br>
However, it’s possible to add a condition:</p>
<div class="md-table">
<table>
<thead>
<tr>
<th>Event</th>
<th>Origin</th>
<th>Fork</th>
</tr>
</thead>
<tbody>
<tr>
<td>pull_request</td>
<td>:heavy_check_mark:</td>
<td>:x:</td>
</tr>
<tr>
<td>pull_request_target</td>
<td>:x:</td>
<td>:heavy_check_mark:</td>
</tr>
</tbody>
</table>
</div><pre><code>jobs:
  job1:
    if: &gt;-
      (
        github.event_name == 'pull_request' &amp;&amp;
        github.event.pull_request.head.repo.full_name == github.repository
      ) || (
        github.event_name == 'pull_request_target' &amp;&amp;
        github.event.pull_request.head.repo.full_name != github.repository
      )
</code></pre>
<p>The undesired combinations are skipped. The other two have access to secrets. If the PR branch is in the same repo, then changes to the workflow contained in the PR will apply immediately (i.e. you as maintainer can test workflow changes). If the PR has its branch in a forked repo, then workflow changes won’t be run (the origin version of the workflow is used). In the skipped combination of <code>pull_request</code> and fork, workflow changes would be taken into account, but since there is no access to secrets, external users will not be able to leak them unless you merge malicious changes.</p>
